docs/general/developer-resources/sapphire-paratime/writing-dapps-on-sapphire.md

```diff
@@ -0,0 +1,183 @@
+
+## Writing Secure DApps
+
+### Wallets
+
+Sapphire is compatible with popular self-custodial wallets including MetaMask,
+Ledger, Gnosis Safe, Brave, and so forth. You can also use libraries like Web3.js
+and Ethers to create programmatic wallets. In general, if it generates secp256k1
+signatures, it'll work just fine.
+
+### Languages & Frameworks
+
+Sapphire is programmable using any language that targets the EVM, such as Solidity
+and Vyper. If you prefer to use an Ethereum framework like Hardhat or Truffle, you
+can also use those with Sapphire; all you need to do is set your Web3 gateway URL.
+You can find the details of the Oasis Sapphire Web3 gateway
+[here](/general/developer-resources/sapphire-paratime#web3-gateway).
+
+
+### Transactions & Calls
+
+Transactions and calls must be encrypted and signed for maximum security. You can
+use the [`@oasislabs/sapphire-paratime`] JS package to make your life easy. It'll
+handle cryptography and signing for you.
+
+One thing to keep in mind is that `msg.sender` will be zeroed for calls that are
+not signed. If you want to use `msg.sender` for access control, the call must be
+signed. If you want to avoid signature popups in the user's wallet, just set the
+`from` address to all zeros. The JS library will do this for you.
+
+### Contract State
+
+The Sapphire state model is like Ethereum's except for all state being encrypted
+and not accessible to anyone except the contract. The contract, executing in an
+active (attested) Oasis compute node is the only entity that can request its its
+state encryption key from the Oasis key manager. Both the keys and values of the
+items stored in state are encrypted, but the size of either is *not* hidden. You
+app may need to pad state items to a constant length, or use other obfuscation.
+Oasis storage nodes may also infer computation based on storage access patterns,
+so you may need to obfuscate that, too.
+
+Contract state may be made available to third parties through logs/events, or
+explicit getters.
+
+### Contract Logs
+
+Contract logs/events (e.g., those emitted by the Solidity `emit` keyword)
+are exactly like Ethereum. Data contained in events is *not* encrypted.
+Precompiled contracts are available to help you encrypt data that you can
+then pack into an event, however.
+
+:::danger Unmodified contracts may leak state through logs
+
+Base contracts like those provided by OpenZeppelin often emit logs containing
+private information. If you don't know they're doing that, you might undermine
+the confidentiality of your state. As a concrete example, the ERC-20 spec
+requires implementers to emit an `event Transfer(from, to, amount)`, which is
+obviously problematic if you're writing a confidential token. What you can
+do instead is fork that contract and remove the offending emissions.
+
+:::
+
+
+### Precompiles
+
+Sapphire provides a few extra precompiled contracts that you may find useful.
```

Comment:
These precompiles aren't available yet, right?

Reply:
not yet. I'll make a note of that